mwlwifi: fix 6.1 Kernel support

stklcode opened the pull request:
Remove `force_uaccess_*` API usage and replace calls from _pci-dma-compat.h_ for Kernel >= 5.18, so that the module compiles against Kernel 6.1

Build and run tested on a Linksys WRT1900ACS with two 88W8864 chips.

The patches are not really pretty, maybe I can figure out a more elegant solution, but there are already some switches by kernel version, so I just added another one.

stklcode committed:
mwlwifi: fix 6.1 Kernel support

Remove force_uaccess_* API usage.
Replace calls from pci-dma-compat.h

Signed-off-by: Stefan Kalscheuer <stefan@stklcode.de>

Ansuel commented:
Did we tried to send this upstream to mwlwifi? Maybe there is a chance it gets merged there?

stklcode commented:
I will definitely send it upstream. Just want to take some additional time to maybe find a cleaner solution.
The proposed patch enabled me to properly test Kernel 6.1 on my WRT1900 test router for now.

Feel free to ignore it until an upstream attempt was made, otherwise we can just remove it afterwards.

Ansuel commented:
the first patch have unintended change (tab converted to whitespace)
the second patch can be pushed upstream and you can just drop the pci stuff and use dma api directly. I checked back in time and the api were legacy even before 4.10 o.O

stklcode commented:
> the second patch can be pushed upstream and you can just drop the pci stuff and use dma api directly. I checked back in time and the api were legacy even before 4.10 o.O

That's what I was hoping for, but didn't check yet. Thanks for that info, definitely better than the switches.

hauke commented:
Does this still work with kernel 5.15?

hauke commented on package/kernel/mwlwifi/patches/006-remove-uaccess-and-get_fs-calls-from-PCIe-for-Kenel-.patch:
```diff
@@ -0,0 +1,49 @@
+ 	set_fs(oldfs);
+-#else
++#elif LINUX_VERSION_CODE < KERNEL_VERSION(5,18,0)
+ 	force_uaccess_end(oldfs);
```
Is calling `force_uaccess_begin` and `force_uaccess_end` not needed at all any more?

stklcode replied:
`force_uaccess_{begin,end}` were introduced for transitional purpose to get rid of `get_fs()` calls in 5.10 (https://github.com/torvalds/linux/commit/3d13f313ce4c34c524ccc37986fe77172f601ff3) followed by the introduction of `CONFIG_SET_FS` to migrate one arch after another (https://github.com/torvalds/linux/commit/5e6e9852d6f76e01b2e6803c74258afa5b432bc5)

The internal read/write logic has been refactored, s.t. driver code can transparently operate on userspace/kernel pointers without messing with boundaries individually (mainly https://github.com/torvalds/linux/commit/4d03e3cc59828c82ee89ea6e27a2f3cdf95aaadf and https://github.com/torvalds/linux/commit/36e2c7421f02a22f71c9283e55fdb672a9eb58e7).

Finally, `CONFIG_SET_FS` was removed in 5.18 for all architectures ([ChangeLog-5.18](https://cdn.kernel.org/pub/linux/kernel/v5.x/ChangeLog-5.18)) and the entire `{g,s}et_fs()` and `force_uaccess_{begin,end}` logic removed in https://github.com/torvalds/linux/commit/967747bbc084b93b54e66f9047d342232314cd25

(some additional reading on the evolution of `set_fs()` can be found here: https://lwn.net/Articles/832121/)

----

So, if I didn't miss anything, the calls should not require any replacement.
I'll try to add some compact and hopefully meaningful wording to the patch description.

stklcode commented:
> Does this still work with kernel 5.15?

Yes. For 5.15 the effective code is equivalent.

The `pci_*` methods are just convenience wrappers around `dma_*` logic, see [include/linux/pci-dma-compat.h (v5.15)](https://github.com/torvalds/linux/blob/v5.15/include/linux/pci-dma-compat.h)

e.g.
```c
static inline int pci_set_dma_mask(struct pci_dev *dev, u64 mask)
{
	return dma_set_mask(&dev->dev, mask);
}
```

That was unified back in 4.6 (https://github.com/torvalds/linux/commit/bc4b024a8b8bd7dceb2697299aad2bda57d065e0) and most of this stuff has been around since the 2.6 days, so IMO it's safe to assume this won't break anything , even if we would use an updated upstream version in 5.10

Build and run testet the code on a Linksys WRT1900ACS with Kernel 5.15.117 and 6.1.34

stklcode commented:
Added some upstream references to both the patch descriptions and rebased on top of the current _main_ branch.

Patches are identical to the pending upstream proposal.

pesa1234 reviewed (commented):
Tested and it works

robimarko commented:
Now that mvebu has 6.1, this PR is really neccessary.

stklcode commented:
To build for all devices with default packages, definitely.

There was not much feedback so far. I hopefully pointed out why I think the proposed changes should be correct (thanks for the questions to clear this up) and there have been some positive test feedbacks (thanks for testing).

An upstream solution should be preferred imho (linked PR above), so we only have to update the package, but we can also patch it for now and drop the patch series later, if that process takes too much time. There’s only not that much movement in the upstream sources, so I guess the overhead should also be acceptable.

robimarko commented:
Upstream is basically dead except for then suddenly merging PR-s once a year.
So, until it gets merged there we should carry this locally.

panchomira commented:
Im waiting for this
